File: src/main/java/io/zeitwert/fm/search/adapter/api/rest/SearchController.java

package io.zeitwert.fm.search.adapter.api.rest;

import java.util.Collections;
import java.util.List;

import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import io.dddrive.search.model.SearchResult;
import io.dddrive.search.service.api.SearchService;
import io.zeitwert.fm.search.adapter.api.rest.dto.SearchResultDto;

@RestController("searchController")
@RequestMapping("/rest/search")
public class SearchController {

	private static final Integer SEARCH_RESULT_SIZE = 10;

	private final SearchService searchService;

	SearchController(SearchService searchService) {
		this.searchService = searchService;
	}

	@GetMapping()
	public ResponseEntity<List<SearchResultDto>> find(@RequestParam String searchText,
			@RequestParam(required = false) List<String> itemTypes) {
		List<SearchResult> items = this.searchService.find(itemTypes, searchText, SEARCH_RESULT_SIZE);
		Collections.sort(items, Collections.reverseOrder());
		return ResponseEntity.ok(
				items.stream().limit(SEARCH_RESULT_SIZE).map(sr -> SearchResultDto.fromItem(sr)).toList());
	}

}

File: fm-domain/src/main/java/io/zeitwert/fm/search/api/rest/SearchController.kt
package io.zeitwert.fm.search.api.rest

import io.zeitwert.app.model.SessionContext
import io.zeitwert.dddrive.api.jsonapi.dto.EnumeratedDto
import io.zeitwert.dddrive.ddd.model.SearchResult
import io.zeitwert.dddrive.ddd.service.api.SearchService
import io.zeitwert.fm.search.api.rest.dto.SearchResultDto
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController
import java.util.*

@RestController("searchController")
@RequestMapping("/rest/search")
class SearchController {

	@Autowired
	lateinit var searchService: SearchService

	@Autowired
	lateinit var sessionContext: SessionContext

	@GetMapping
	fun find(
		@RequestParam searchText: String,
		@RequestParam(required = false) itemTypes: List<String>,
	): ResponseEntity<List<SearchResultDto>> {
		val items: List<SearchResult> = searchService.find(sessionContext, itemTypes, searchText, SEARCH_RESULT_SIZE)
		Collections.sort(items, Collections.reverseOrder())
		return ResponseEntity.ok(
			items
				.stream()
				.limit(SEARCH_RESULT_SIZE.toLong())
				.map {
					SearchResultDto(
						tenantId = it.tenantId,
						itemType = EnumeratedDto.of(it.aggregateType)!!,
						id = it.id.toString(),
						caption = it.caption,
						rank = it.rank,
					)
				}.toList(),
		)
	}

	companion object {

		private const val SEARCH_RESULT_SIZE = 10
	}

}
